src/server.c

```diff
@@ -5214,6 +5224,13 @@ void closeClildUnusedResourceAfterFork() {
 
 /* purpose is one of CHILD_TYPE_ types */
 int redisFork(int purpose) {
+    if (isMutualExChildType(purpose)) {
+        if (hasActiveChildProcess())
+            return -1;
+
+        openChildInfoPipe();
+    }
```

Comment:
Till now, we have checked if there are current active child processes before call `redisFork`, so i think we can simplify this code, and respect your code below
```suggestion
    if (isMutualExChildType(purpose)) openChildInfoPipe();
```

Reply:
i think that for safety, we better keep this one here too.
in the other places we add that check before calling `redisFork` in order to avoid doing other excess work (like crating a pipe and then destroying it)

Reply:
Hi @oranagra maybe i miss some things. I think we call `hasActiveChildProcess` before `redisFork` that guarantee there is no child process. before this refactor, we also call `redisFork ` directly. Yes, it is safe, but a little repeated.

@YaacovHazan I am sorry, i forget to submit my other comment, for the function `isMutualExChildType`, i think `MutualEx` is a little unclear, how about `isMutexChildType`

Reply:
@ShooterIT initially Yaacov removed these calls to `hasActiveChildProcess` from before calling fork since the fork method already handles that. but i asked him to add a few (or maybe all) of them back since in some of these the check for `hasActiveChildProcess` was a big earlier, thus avoiding some extra work (like creating a pipe) if there's a child already active.

but that doesn't mean that in the future someone may want to just call `redisFork` without doing that check, so i think we should keep it just for safety (it doesn't have any cost other than an additional 2 lines or so).

regarding the name, `isMutualExChildType` refers to "mutually exclusive". i know that "Mutex" also refers to it. but being a programmer for so many years, the word Mutex in my head refers to the multi-threading synchronization primitive, and not a short for "Mutually Elusive".
Point is, that to me, your suggestion is not attractive. (i agree the current name is cumbersome, but i don't think making it shorter improves it.